Cleaning up CMakeLists

henryiii opened the pull request:
Also makes it easier to include in a project that wants to use Hydra.

There really are very few changes in this PR, mostly cleanup. The only things that are not cleanup changes are:

* CMake 3.1 required instead of 3.0. 3.0 was almost completely unused, and 3.1 has much better defaults. ROOT requires 3.4.3+, so upping that to 3.4 wouldn't be a bad idea either.
* Output verbosity is decreased a bit. It's hard to see real errors and warnings, and it makes a new user think that something is wrong. You can still get verbose build output by running `make VERBOSE=1`
* Examples only build by default if this is the main CMakeLists. Projects that include Hydra as a subpackage will not have the examples build by default (though `-DHYDRA_EXAMPLES=ON` or `OFF` works, too.

The other cleanup changes are fixing indentation and normalizing upper/lower case. There were several typos that were corrected, as well.

Please cherry-pick this commit into Develop, as well.

Built with all four backends on Goofy to verify that I didn't inadvertently break anything. I'll post a few inline comments on the PR.

henryiii commented on CMakeLists.txt:
```diff
@@ -1,155 +1,143 @@
 ### CMakeList for MC Hydra
 
-cmake_minimum_required(VERSION 3.0)
-cmake_policy(SET CMP0054 OLD)
```
Do not set a policy to OLD unless you have to! CMake will drop the old versions of policies at some point. (I think this one is being dropped in 3.11). This is for quoted if statements and the old behavior is very confusing for most users. I did not see any remaining if statements that would be affected by this.

AAAlvesJr replied:
Did you tested what happens removing this ?

henryiii replied:
Yes, I built all four backends on Goofy with this PR.

AAAlvesJr replied:
If I remove the line -cmake_policy(SET CMP0054 OLD) , I get this from cmake:

```
 CMake Warning (dev) at /usr/share/cmake/Modules/FindOpenMP.cmake:200 (if):
  Policy CMP0054 is not set: Only interpret if() arguments as variables or
  keywords when unquoted.  Run "cmake --help-policy CMP0054" for policy
  details.  Use the cmake_policy command to set the policy and suppress this
  warning.

  Quoted variables like "c" will no longer be dereferenced when the policy is
  set to NEW.  Since the policy is not set the OLD behavior will be used.
Call Stack (most recent call first):
  /usr/share/cmake/Modules/FindOpenMP.cmake:324 (_OPENMP_GET_FLAGS)
  CMakeLists.txt:70 (find_package)
This warning is for project developers.  Use -Wno-dev to suppress it.
 ```

How do you interpret this?

henryiii replied:
That is because you set the minimum version to 3.0. You should set it to 3.1. (3.0 was not used heavily, there were lots of fixes in 3.1)

AAAlvesJr replied:
So, please promote it to 3.1.

henryiii replied:
It is 3.1 in this PR. :)

henryiii commented on CMakeLists.txt:
```diff
@@ -1,155 +1,143 @@
-# The version number.
-set (Hydra_VERSION_MAJOR 1)
-set (Hydra_VERSION_MINOR 0)
-set (Hydra_VERSION_PATCH 0)
```
Shouldn't this be 2.0.0? These names are auto-generated if you set the VERSION keyword in project.

AAAlvesJr replied:
Ideally, should pick-up the correct version from Hydra.h as doxygen does.  Can you implement this?

henryiii replied:
Sure, will be able to finish this later today.

henryiii commented on CMakeLists.txt:
```diff
@@ -1,155 +1,143 @@
- MESSAGE(STATUS "===> Compiling with Intel")
-elseif ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "MSVC")
- MESSAGE(STATUS "===> Compiling with MSVC")
-endif()
```
This is already printed by CMake's configure.

AAAlvesJr replied:
Sure. I will just comment this out.  A piece of code  to place conditional commands depending the compiler.

henryiii commented on CMakeLists.txt:
```diff
@@ -1,155 +1,143 @@
-# warn user if system is not UNIX
+# Warn user if system is not UNIX
 if(NOT UNIX)
-  message(FATAL_ERROR "This is an unsupported system.")
```
Warn about a system not being supported, never kill it. Someone may add support for a new system down the road! (And, even the comment says it will be a warning)

AAAlvesJr replied:
This is a matter of opinion. The person implementing support for another system can perfectly change it. I wll think about....

henryiii replied:
I know it's a matter of opinion, but you said "Warn the user" so I set to to warn. Please tell me if you want it to be a fatal error, and I'll change it, so you can accept the PR (you can also push to this branch if you want changes).

AAAlvesJr replied:
I would like to keep this like that for the moment. Please, can you change it?

eduardo-rodrigues replied:
Augusto, why do you want to make life complicated to the random user who would like to try it on some other architecture? That's it's problem in the end, no? Why would you even care and make it impossible with FATAL?

AAAlvesJr replied:
Hi Eduardo, Hydra is not supported for other platforms than linux.  I make a huge deal of effort to keep the code as agnostic as possible, but more testing is necessary before to add support for other systems. 

Build is just required to compile and run the examples. I tested the examples in linuxes machines, I don't know how they would behave in other systems. User who hacked it and ran on other systems are encouraged to report their experience in the hydra-library-users google group. 

Henry, please keep Fatal_error.

eduardo-rodrigues replied:
I am not saying above that you should give support for other platforms. I am saying you should not close the door and make it impossible or really hard for anyone to even try the package in a platform you have not tested. That's all. Having to hack for even trying to build is not great IMHO. Your call.

AAAlvesJr replied:
"... really hard for anyone to even try the package in a platform you have not tested."
This comment is misleading Eduardo, building is just necessary for the examples ...
It is a header only library.  Nothing prevent users from other platforms of include the code in their source trees.

eduardo-rodrigues replied:
I rest my case. Personally, the first thing I try with a new package is to build it and/or run some examples. If I cannot get it to work on my system that's frustrating from the onset. Maybe that's just me. Fine.

henryiii replied:
Why would someone in approach  not try to hack the package? You left out that in both cases you may have to hack Hydra to get it to work, the approach 1 just means you hack 1 less line. (not everyone is great at reading CMake, by the way). Not immediately killing it in a FATAL_ERROR would help them see what needs to be hacked immediately on their system and help them start fixing things.

Remember, this will *only* make a difference for someone brave enough to try it on a different system (FreeBSD, Windows, etc.). And in either approach, they *still get the unsupported warning*.

The only benefit I see to Approach 1 is if the only required change is the one single line, then you will learn that it really does work on that system, vs. a small chance that someone in Approach 2 won't bother to report that the warning is not needed on their system if it just works.

I'm 100% okay with either approach, though I do like Approach 2 a bit better. I *think* the result of this thread was to keep FATAL_ERROR, so I'll do that and let Augusto change it if he decides to go with Approach 2.

henryiii replied:
I muddled approaches 1 and 2, corrected above.

henryiii commented on CMakeLists.txt:
```diff
@@ -1,155 +1,143 @@
 
-SET(Hydra_CMAKE_DIR "${PROJECT_SOURCE_DIR}/cmake")
-SET(CMAKE_MODULE_PATH "${Hydra_CMAKE_DIR}" ${CMAKE_MODULE_PATH})
-SET(CMAKE_VERBOSE_MAKEFILE  ON)
```
`make VERBOSE=1` does the same thing

AAAlvesJr replied:
(yes)

henryiii commented on CMakeLists.txt:
```diff
@@ -1,155 +1,143 @@
-CHECK_CXX_COMPILER_FLAG("--std=c++11" COMPILER_SUPPORTS_CXX11)
+check_cxx_compiler_flag("--std=c++11" COMPILER_SUPPORTS_CXX11)
 if(NOT COMPILER_SUPPORTS_CXX11)
- message(FATAL "The compiler ${CMAKE_CXX_COMPILER} has no C++11 support. Please use a different C++ compiler.")
```
FATAL_ERROR, not FATAL

AAAlvesJr replied:
(yes)

henryiii commented on CMakeLists.txt:
```diff
@@ -158,31 +146,34 @@ file(MAKE_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/plots/dalitz)
-add_subdirectory(examples/histograming)
-add_subdirectory(examples/async)
-add_subdirectory(examples/misc)
-add_subdirectory(examples/phys)
```
These should be protected with an option, makes including as part of another project easier (hint hint :) )

AAAlvesJr replied:
In which context these files would be included in another project? I just dont see your point here. The project is header-only, there is nothing to compile and these are only example...

henryiii replied:
You can include it inside another project, and then `add_subdirectory` on it. This would then add the CMake targets, settings, compiler flags, and functions from Hydra instead of the user being responsible for setting them him/herself. For example, that's supported with PyBind11.

PS: It is much more useful with CMake 3-style targets, as a future improvement.

AAAlvesJr replied:
I am not sure if i understand what you mean. If it was to add a flag here, it should be set default to ON.  For include hydra in another projects, I think it is much better have something like a FindHydra.cmake...

henryiii replied:
I did set the default to ON, unless including in another package. This is how PyBind11, fmtlib, and many other (header only) libraries work. However, you will see it done this way:

```
# Determine if fmt is built as a subproject (using add_subdirectory)
# or if it is the master project.
set(MASTER_PROJECT OFF)
if (CMAKE_CURRENT_SOURCE_DIR STREQUAL CMAKE_SOURCE_DIR)
  set(MASTER_PROJECT ON)
endif ()
```

Would that be fine?

AAAlvesJr replied:
Seems okay to me, but I am not familiar with this mechanism. I will check better later. Please add a comment.

henryiii replied:
It's in the latest commit, and has a comment on what it does. :)

henryiii commented on CMakeLists.txt:
```diff
@@ -193,11 +184,10 @@ install(FILES
 if(BUILD_DOCUMENTATION)
 
     if(NOT DOXYGEN_FOUND)
-        message(FALTAL_ERROR "Doxygen is needed to build the documentation.")
```
FATAL_ERROR, not FALTAL_ERROR. I'm guessing you have always had Doxygen installed.

AAAlvesJr replied:
Not at all. The documentation processing needs be explicitly requested passing   -DBUILD_DOCUMENTATION=ON 
to cmake command. I will correct the typo.

henryiii replied:
I meant you always had Doxygen installed when you passed `-DBUILD_DOCUMENTATION=ON`

AAAlvesJr commented:
The version of cmake required in hydra has nothing to do with the version required in ROOT, because hydra examples compilation will not try to install ROOT etc.  The only examples that requires ROOT to compile and run are the fit related stuff. So, I think keeping lower cmake version is not a bad idea. 

Hydra is header only. Projects using hydra will add the hydra to the includes and so on...

henryiii commented:
I've fixed the issues listed above, please review again. I've added the logic from your FindHydra (I have a better FindHydra that I'll send a link to in a few days), but is the version number 2.0.1 or 2.2.1? Also, just to let you know, your FindHydra version finding is broken, the variable is `"Hydra_VERSION"`, not `"HYDRA_VERSION"`.

AAAlvesJr commented:
Thank for watching my work so closely. I  still need test the FindHydra.cmake. I am satisfied with the functionality it provides though.  There are major changes on develop that will be merged on rc_2.0 before the release. So maybe you want to await a feel days to conclude this cmake p.r.

henryiii commented:
By the way, my "better" FindHydra is not intended to replace yours, since it will require a very new version of CMake. Even though Hydra is header only, setting up the multiple backend targets, compile time definitions, required defines, etc. is much more work than the normal user will want to do (especially in CMake). My FindHydra (or including this as a package) would provide nice tools for a user to use Hydra easily (2-5 lines of CMake).

I would recommend merging this PR soon, then making whatever changes you need to the modified CMakeLists.txt instead of working in parallel. I'm happy to change the merge target to develop if that's better, though.

AAAlvesJr commented:
Please, let's stop to speak in the name of the "normal user", citing things like "what the normal user would do" or "wouldn't  do", "how he thinks" etc... as arguments to push ahead PRs here. There is no serious research to support these claims. There are only prejudices. 

Let me cite one example: Thrust library. The user base of Thrust is vast, it has many active contributors and a vibrating user community. On the other hand there are no such fancy-automatic mechanisms there. Instead, Thrust is well coded and documented. 

This is the key: good documentation. 

So, going back to the possibility to distribute with Hydra cmake macros to set back-ends and other things automatically. First, personally speaking, I am quite opposite to the intensive usage of such automatic mechanisms because they tend to promote the idea that packages as black-boxes. Even if  this can work well for certain packages,  I would like to keep Hydra out of this trend. Second, the combination between compiler flags, back-ends, etc. can be very specific, so I dont want to find myself maintaining such stuff in a almost per user basis, because the package itself created such expectations on the users. I plan to invest the time developing and maintaining documentation, to give the users the best conditions to take their own decisions when using the package. 

I will consider to accept your current PR after the merge, sorry.

henryiii commented:
This PR has nothing to do with "FindHydra" at all. I just said that the feature I added *that you asked for*, the version discovery, was based on your FindHydra. That's all.

I actually explicitly said my FindHydra script would **not** replace the one inside Hydra.

And in the develop branch, why did you move from "THRUST_*" variables to "HYDRA_*" variables for setting the backends? If a Thrust user comes to Hydra, they would expect the former, which is well documented, rather than the latter.

AAAlvesJr commented:
I understood you PR has nothing to do with FindHydra.

Hydra and his built-in Thrust version should not conflict in anyway with the Thrust installed in the user's system. Hydra will set his back-ends using different macros. It allows mix hydra with thrust in the same programs and control each of them independently.

henryiii commented:
Resolved conflicts. Also am now targeting develop, since that way you can merge into develop, test and modify if needed, then make this a part of the PRs to 2.0rc. (Let me know if this is not ideal)
